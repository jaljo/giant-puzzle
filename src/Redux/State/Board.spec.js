import {
  findTileWithCharacter,
  findTileByCoordinates,
} from './../../Util'
import {
  INITIAL_STATE,
  arrowKeyPressed,
  clear,
  default as reducer,
  destinationTileFound,
  moveCharacter,
} from './Board'
import {
  MAIN_CHARACTER,
  GUARDIAN_REGULAR,
  GUARDIAN_REVERSE,
} from './Characters'

describe('Redux :: State :: Board', () => {
  it('reduces to initial state by default', () => {
    expect(reducer()).toEqual(INITIAL_STATE)
    expect(INITIAL_STATE).toMatchSnapshot()
  })

  it('reduces arrowKeyPressed action', () => {
    const s1 = reducer(INITIAL_STATE, arrowKeyPressed('right'))

    expect(
      findTileWithCharacter(MAIN_CHARACTER.id)(s1).char.direction
    ).toBe('right')
    expect(
      findTileWithCharacter(GUARDIAN_REGULAR.id)(s1).char.direction
    ).toBe('right')
    expect(
      findTileWithCharacter(GUARDIAN_REVERSE.id)(s1).char.direction
    ).toBe('left')
  })

  it('reduces moveCharacter action', () => {
    const mockDestTile = { x: 0, y: 0, char: null }
    const dtf = destinationTileFound(MAIN_CHARACTER.id, 'up', mockDestTile)
    const s1 = reducer(INITIAL_STATE, moveCharacter(dtf))

    // the character should have been moved to the destination tile
    expect(
      findTileWithCharacter(MAIN_CHARACTER.id)(s1)
    ).toMatchObject({
      x: 0,
      y: 0,
    })

    // the character should have been removed from its original place
    expect(
      findTileByCoordinates(2, 2)(s1)
    ).toMatchObject({
      x: 2,
      y: 2,
      char: null,
    })

    // other characters should not have moved
    expect(
      findTileWithCharacter(GUARDIAN_REGULAR.id)(s1)
    ).toEqual(
      findTileWithCharacter(GUARDIAN_REGULAR.id)(INITIAL_STATE)
    )
    expect(
      findTileWithCharacter(GUARDIAN_REVERSE.id)(s1)
    ).toEqual(
      findTileWithCharacter(GUARDIAN_REVERSE.id)(INITIAL_STATE)
    )
  })

  it('reduces clear action', () => {
    expect(reducer([], clear())).toEqual(INITIAL_STATE)
  })
})
