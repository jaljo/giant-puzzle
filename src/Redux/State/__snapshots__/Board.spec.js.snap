// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux :: State :: Board reduces to initial state by default 1`] = `
Array [
  Array [
    Object {
      "char": null,
      "locked": false,
      "x": 0,
      "y": 5,
    },
    Object {
      "char": Object {
        "asset": "fox",
        "direction": "down",
        "id": "guardian-reverse",
      },
      "locked": false,
      "x": 1,
      "y": 5,
    },
    Object {
      "char": null,
      "locked": true,
      "x": 2,
      "y": 5,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 3,
      "y": 5,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 4,
      "y": 5,
    },
  ],
  Array [
    Object {
      "char": null,
      "locked": false,
      "x": 0,
      "y": 4,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 1,
      "y": 4,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 2,
      "y": 4,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 3,
      "y": 4,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 4,
      "y": 4,
    },
  ],
  Array [
    Object {
      "char": null,
      "locked": false,
      "x": 0,
      "y": 3,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 1,
      "y": 3,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 2,
      "y": 3,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 3,
      "y": 3,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 4,
      "y": 3,
    },
  ],
  Array [
    Object {
      "char": null,
      "locked": true,
      "x": 0,
      "y": 2,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 1,
      "y": 2,
    },
    Object {
      "char": Object {
        "asset": "chick",
        "direction": "up",
        "id": "main-character",
      },
      "locked": false,
      "x": 2,
      "y": 2,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 3,
      "y": 2,
    },
    Object {
      "char": null,
      "locked": true,
      "x": 4,
      "y": 2,
    },
  ],
  Array [
    Object {
      "char": null,
      "locked": true,
      "x": 0,
      "y": 1,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 1,
      "y": 1,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 2,
      "y": 1,
    },
    Object {
      "char": null,
      "locked": false,
      "x": 3,
      "y": 1,
    },
    Object {
      "char": null,
      "locked": true,
      "x": 4,
      "y": 1,
    },
  ],
  Array [
    Object {
      "char": null,
      "locked": true,
      "x": 0,
      "y": 0,
    },
    Object {
      "char": null,
      "locked": true,
      "x": 1,
      "y": 0,
    },
    Object {
      "char": Object {
        "asset": "fox",
        "direction": "up",
        "id": "guardian-regular",
      },
      "locked": false,
      "x": 2,
      "y": 0,
    },
    Object {
      "char": null,
      "locked": true,
      "x": 3,
      "y": 0,
    },
    Object {
      "char": null,
      "locked": true,
      "x": 4,
      "y": 0,
    },
  ],
]
`;
